src/hotspot/share/memory/arena.hpp

```diff
@@ -151,9 +150,11 @@ class Arena : public CHeapObj<mtNone> {
       return old;
     }
   }
-  // Further assume size is padded out to words
+
+  // Allocate in the arena, assuming the size has been aligned to size of pointer, which
+  // is 4 bytes on 32 bits, hence the name.
   void *Amalloc_4(size_t x, AllocFailType alloc_failmode = AllocFailStrategy::EXIT_OOM) {
-    assert( (x&(sizeof(char*)-1)) == 0, "misaligned size" );
+    assert((x & (sizeof(char*)-1)) == 0, "misaligned size");
     debug_only(if (UseMallocOnly) return malloc(x);)
     if (!check_for_overflow(x, "Arena::Amalloc_4", alloc_failmode))
```

Comment:
[pre-existing] Missing brackets for the multiline `if` is contrary to the style guide.  Similarly in Amalloc.

Reply:
[pre-existing: I can't comment directly on line 161, so adding here.]
`if (_hwm + x > _max)` has a risk of overflow if x is excessively large. Better is `if (_max - _hwm < x)`.  Similarly in Amalloc.

Reply:
[pre-existing. I can't comment directly on line 161, so adding here.]
The `if` to range check `x` could be an `else if`.  Similarly in Amalloc.

Reply:
1, 3.  Ok, turned to else if with appropriate brackets.
2. Both of these places have a call check_for_overflow that checks that so rearranging the code leads to casts since _hwm and _max are char* and x is size_t.

Reply:
check_for_overflow does more than just the pointer comparison.  It conditionally calls vm_exit_out_of_memory.  I don't want to copy this into both of the Amalloc function, and it has a nice name vs. squinting and trying to understand why "if (pointer_delta(_max, _hwm, 1) < x)" does the same thing. As is, this is readable and I don't want to change it.

Reply:
Such a use of pointer_delta is one of its canonical use-cases, as discussed in the comment describing that function.

grow does everything that check_for_overflow does, except correctly.  check_for_overflow can incorrectly report an error if the current chunk happens to be near the end of the address space, even though malloc can succeed.

The current code with the call to check_for_overflow is incorrect, slower, and larger than it needs to be.  It's got nothing at all going for it except the "whence" string, which is not enough to make up for the other problems.  If that feature is important, add it to grow().

Reply:
I'll file another RFE to fix it.  I hope you can do it because I would only cut/paste the pointer diff expressions that you type here.  This has already exceeded the original RFE change. JDK-8270217
One backlog item removed, two created JDK-8270217.